proposals/NNNN-cbuffer-contexts.md

```diff
@@ -0,0 +1,93 @@
+<!-- {% raw %} -->
+
+# Refined `cbuffer` Contexts
+
+## Instructions
+
+* Proposal: [NNNN](NNNN-cbuffer-contexts.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz), [Helena Kotas](https://github.com/hekota)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Planned Version: 202x
+* Issues: [DXC #4514](https://github.com/microsoft/DirectXShaderCompiler/issues/4514)
+
+## Introduction
+
+This proposal aligns `cbuffer` declarations more closely with syntactic and
+semantic behaviors that are common in C and C++ to align more closely with user
+expectation and reduce the need for special case handling in the compiler.
+
+## Motivation
+
+The HLSL `cbuffer` declaration is a source of irregularity within the language
+semantics, and the interaction between `cbuffer` declarations and their
+contained declarations can be different and surprising.
+
+In DXC, declarations inside a `cbuffer` are always hoisted out to translation
+unit scope, which effectively ignores the namespace nesting, however the
+declaration context hierarchy within a `cbuffer` is preserved except again in
+the case of a nested `cbuffer` which pops back out to translation unit scope.
+
+This behavior in DXC is clearly a bug, and FXC behaves more reasonably. As an
+example given the following cbuffer declarations:
+
+```hlsl
+namespace ns {
+  cbuffer CB {
+    float a;
+
+    namespace ns2 {
+      cbuffer CB2 {
+        float b;
+      } // cbuffer CB2
+    } // namespace ns2
+  } // cbuffer CB
+} // namespace ns
+```
+[FXC @ ShaderPlayground](https://shader-playground.timjones.io/22396dd8aec3318b3e9a598c612f156e)
+[DXC @ Compiler Explorer](https://godbolt.org/z/WYK7jvvfP)
+
+In DXC, `a` and `b` are referenced directly with no namespace qualifications,
```

Comment:
It is actually possible to reference `a` and `b` in DXC by both qualified and unqualifies names as long as the namespace contains more than just a `cbuffer`: https://godbolt.org/z/Ph57WYexr

Reply:
I feel like this needs a more emotion-packed reaction than I can do just by GitHub's emoji... Something like this wave of emotions:

🤯😲 😱 😭